spec/api-web-contents-spec.js

```diff
@@ -98,6 +98,114 @@ describe('webContents module', function () {
     })
   })
 
+  describe('before-input-event event', () => {
+    it('can prevent document keyboard events', (done) => {
+      w.loadURL('file://' + path.join(__dirname, 'fixtures', 'pages', 'key-events.html'))
+      w.webContents.once('did-finish-load', () => {
+        let error = null
+        const onEvent = (event) => {
+          error = new Error('Unexpected keydown message')
+        }
+        ipcMain.on('keydown', onEvent)
+
+        let beforeInputEventReceived = false
+        w.webContents.on('before-input-event', (event, input) => {
+          assert.equal(input.type, 'keyDown')
+          assert.equal(input.key, 'a')
+          beforeInputEventReceived = true
+          event.preventDefault()
+        })
+
+        w.webContents.sendInputEvent({type: 'keyDown', keyCode: 'a'})
+
+        setTimeout(() => {
+          assert.ok(beforeInputEventReceived)
+          ipcMain.removeListener('keydown', onEvent)
+          done(error)
+        }, 500)
```

Comment:
Yeah, `preventDefault` does not work for events originating from the main process since the main process dispatches them to the listening render processes asynchronously via the remote module and at that point it is too late to prevent them.

You could work around them using a technique like https://github.com/electron/electron/blob/f1f73eff04924fc7fb9838f6a60d7ea4454969a1/spec/static/main.js#L140-L143

Add a specific `ipcMain` channel that sets things up in `spec/main.js` and then you tell it to listen to an event, optionally prevent default, and callback afterwards and verify that from the render process.